miasm2/core/asmblock.py

```diff
@@ -606,45 +722,58 @@ def add_node(self, block):
-                self.add_edge(block, dst, constraint.c_t)
+                self.add_constraint(block.label, dst.label, constraint.c_t)
 
         return status
 
-    def del_node(self, block):
-        super(AsmCFG, self).del_node(block)
-        del self._label2block[block.label]
-
     def merge(self, graph):
         """Merge with @graph, taking in account constraints"""
         # -> add_edge(x, y, constraint)
-        for node in graph._nodes:
-            self.add_node(node)
+        self._loc_key_to_block.update(graph._loc_key_to_block)
+        for label in graph._nodes:
+            self.add_node(label)
+            if label in graph._loc_key_to_block:
+                self.add_block(graph._loc_key_to_block[label])
         for edge in graph._edges:
             # Use "_uniq_" beacause the edge can already exist due to add_node
-            self.add_uniq_edge(*edge, constraint=graph.edges2constraint[edge])
+            self.add_constraint(*edge, constraint=graph.edges2constraint[edge])
+
 
     def node2lines(self, node):
-        yield self.DotCellDescription(text=str(node.label.name),
+        if self.symbol_pool is None:
+            label_name = str(node)
+        else:
+            label_name = self.symbol_pool.str_loc_key(node)
+        yield self.DotCellDescription(text=str(label_name),
                                       attr={'align': 'center',
                                             'colspan': 2,
                                             'bgcolor': 'grey'})
-
-        if isinstance(node, AsmBlockBad):
-            yield [self.DotCellDescription(
-                text=node.ERROR_TYPES.get(node._errno,
-                                          node._errno),
-                                           attr={})]
+        block = self._loc_key_to_block.get(node, None)
+        if node is None:
             raise StopIteration
-        for line in node.lines:
+        if isinstance(block, AsmBlockBad):
+            yield [
+                self.DotCellDescription(
+                    text=block.ERROR_TYPES.get(block._errno,
+                                               block._errno
+                    ),
+                    attr={})
+            ]
+            raise StopIteration
+        for line in block.lines:
             if self._dot_offset:
                 yield [self.DotCellDescription(text="%.8X" % line.offset,
                                                attr={}),
-                       self.DotCellDescription(text=str(line), attr={})]
+                       self.DotCellDescription(text=line.to_string(self.symbol_pool), attr={})]
             else:
-                yield self.DotCellDescription(text=str(line), attr={})
+                yield self.DotCellDescription(text=line.to_string(self.symbol_pool), attr={})
 
     def node_attr(self, node):
-        if isinstance(node, AsmBlockBad):
+        block = self._loc_key_to_block.get(node, None)
+        if node is None:
```

Comment:
You probably are checking `block` against `None`, not `node`
And then, the `if` statement is not needed

Reply:
done